zebra-chain/src/sapling/tests/tree.rs

```diff
@@ -41,3 +51,83 @@ fn incremental_roots() {
         );
     }
 }
+
+#[test]
+fn incremental_roots_with_blocks() -> Result<()> {
+    incremental_roots_with_blocks_for_network(Network::Mainnet)?;
+    incremental_roots_with_blocks_for_network(Network::Testnet)?;
+    Ok(())
+}
+
+fn incremental_roots_with_blocks_for_network(network: Network) -> Result<()> {
+    let (blocks, sapling_roots) = match network {
+        Network::Mainnet => (&*MAINNET_BLOCKS, &*MAINNET_FINAL_SAPLING_ROOTS),
+        Network::Testnet => (&*TESTNET_BLOCKS, &*TESTNET_FINAL_SAPLING_ROOTS),
+    };
+    let height = NetworkUpgrade::Sapling
+        .activation_height(network)
+        .unwrap()
+        .0;
+
+    // Load Block 0 (activation block of the given network upgrade)
+    let block0 = Arc::new(
+        blocks
+            .get(&height)
+            .expect("test vector exists")
+            .zcash_deserialize_into::<Block>()
+            .expect("block is structurally valid"),
+    );
+    // Build initial MMR tree with only Block 0
+    let sapling_root0 =
+        sapling::tree::Root(**sapling_roots.get(&height).expect("test vector exists"));
+
+    let mut tree = sapling::tree::NoteCommitmentTree::default();
+
+    // Add note commitments in Block 0 to the tree
+    for transaction in block0.transactions.iter() {
+        for sapling_note_commitment in transaction.sapling_note_commitments() {
+            tree.append(*sapling_note_commitment)
+                .expect("test vector is correct");
+        }
+    }
+
+    // Check if root is correct
+    assert_eq!(sapling_root0, tree.root());
+
+    // Load Block 1 (activation + 1)
+    let block1 = Arc::new(
+        blocks
+            .get(&(height + 1))
+            .expect("test vector exists")
+            .zcash_deserialize_into::<Block>()
+            .expect("block is structurally valid"),
+    );
+    let sapling_root1 = sapling::tree::Root(
+        **sapling_roots
+            .get(&(height + 1))
+            .expect("test vector exists"),
+    );
+
+    // Add note commitments in Block 1 to the tree
+    let mut appended_count = 0;
+    for transaction in block1.transactions.iter() {
+        for sapling_note_commitment in transaction.sapling_note_commitments() {
+            tree.append(*sapling_note_commitment)
+                .expect("test vector is correct");
+            appended_count += 1;
+        }
+    }
+    // We also want to make sure that sapling_note_commitments() is returning
+    // the commitments in the right order. But this will only be actually tested
+    // if there are more than one note commitment in a block.
+    // In the test vectors this applies only for the block 1 in mainnet,
+    // so we make this explicit in this assert.
+    if network == Network::Mainnet {
+        assert!(appended_count > 1);
+    }
+
+    // Check if root is correct
+    assert_eq!(sapling_root1, tree.root());
```

Comment:
```suggestion
    // Load Sapling activation block
    let sapling_activation_block = Arc::new(
        blocks
            .get(&height)
            .expect("test vector exists")
            .zcash_deserialize_into::<Block>()
            .expect("block is structurally valid"),
    );
    // Build initial MMR tree with only the Sapling activation block
    let sapling_activation_block_root =
        sapling::tree::Root(**sapling_roots.get(&height).expect("test vector exists"));

    let mut tree = sapling::tree::NoteCommitmentTree::default();

    // Add note commitments in the Sapling activation block to the tree
    for transaction in sapling_activation_block.transactions.iter() {
        for sapling_note_commitment in transaction.sapling_note_commitments() {
            tree.append(*sapling_note_commitment)
                .expect("test vector is correct");
        }
    }

    // Check if root is correct
    assert_eq!(sapling_activation_block, tree.root());

    // Load the block immediately after Sapling activation (activation + 1)
    let block_after_sapling_activation = Arc::new(
        blocks
            .get(&(height + 1))
            .expect("test vector exists")
            .zcash_deserialize_into::<Block>()
            .expect("block is structurally valid"),
    );
    let block_after_sapling_activation_root = sapling::tree::Root(
        **sapling_roots
            .get(&(height + 1))
            .expect("test vector exists"),
    );

    // Add note commitments in Block 1 to the tree
    let mut appended_count = 0;
    for transaction in block_after_sapling_activation.transactions.iter() {
        for sapling_note_commitment in transaction.sapling_note_commitments() {
            tree.append(*sapling_note_commitment)
                .expect("test vector is correct");
            appended_count += 1;
        }
    }
    // We also want to make sure that sapling_note_commitments() is returning
    // the commitments in the right order. But this will only be actually tested
    // if there are more than one note commitment in a block.
    // In the test vectors this applies only for the block 1 in mainnet,
    // so we make this explicit in this assert.
    if network == Network::Mainnet {
        assert!(appended_count > 1);
    }

    // Check if root is correct
    assert_eq!(block_after_sapling_activation_root, tree.root());
```

Reply:
Oops! I worked on a fix before seeing your comment. I went with `first_block` and `second_block` though looking back now maybe that's not really different. I'll change to use your suggestion.

Reply:
Done in 637f86d5

Reply:
lol this actually went through! i thought i lost it. Your commit looks good!